src/garage/envs/environment.py

```diff
@@ -0,0 +1,188 @@
+"""Base Garage Environment API."""
+
+import abc
+
+
+class Environment(abc.ABC):
+    """The main API for garage environments.
+
+    The main public API methods are:
+        * step()
+        * reset()
+        * render()
+        * visualize()
+        * close()
+
+    Set the following attributes:
+        * action_space: The action space specification
+        * observation_space: The observation space specification
+        * spec: The environment specifications
+        * render_modes: The list of supported render modes
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__getstate__()` and
+        `__setstate()__` to add your custom pickle logic.
```

Comment:
`__setstate__()` and `__getstate__()`

Reply:
you might also mention this EzPickle example as a lightweight way of achieving the pickle functionality most environment implementers want: https://github.com/openai/gym/blob/master/gym/utils/ezpickle.py